File: sdk/python/clearcote/_profileauto.py
"""``profile="auto"`` — resolve a persona from the best source available.

Port of the Node SDK's ``profileauto.ts``.

ORDER: the licensed clearcote service first, a local imported directory as backup.

The service is preferred because its corpus is fresher, larger and curated. But it can be
unreachable for reasons that are none of the caller's fault — no license, an exhausted daily
budget, a rate limit, a network blip, an outage — and in every one of those a local profile is
enormously better than no persona. So the fallback is by design, not a patch.

THE FALLBACK IS ALWAYS ANNOUNCED. A silent downgrade would leave a caller believing they are on a
fresh service profile while they are on a stale local one, and this product is about knowing
exactly what identity you present. ``source`` is on every result and a one-line reason goes to
stderr unless ``quiet``.
"""

from __future__ import annotations

import json
import os
import sys
from typing import Any, Optional

from ._profileimport import import_directory, load_imported_profile
from ._profilelib import select_profile
from ._profilesource import fetch_index, fetch_profile, profile_cache_dir

#: Where an imported (e.g. converted chrome-fingerprints) directory is looked for by default.
DEFAULT_LOCAL_DIR = os.environ.get("CLEARCOTE_PROFILE_SOURCE") or os.path.join(
    os.path.expanduser("~"), ".clearcote", "profiles", "imported"
)


def local_setup_hint(directory: str) -> str:
    """How to populate a local backup directory.

    OPTIONAL, AND NOT SOMETHING WE SHIP. Clearcote redistributes no third-party fingerprints; the
    dataset is fetched by the user from its own repository under its own terms, and we only read
    the converted format. Any directory of clearcote-profile JSON works just as well, which is
    why this is phrased as one example rather than the required path.
    """
    return "\n".join([
        'A local profile directory is optional, and gives "auto" a backup when the service is',
        "unavailable. Any folder of clearcote-profile JSON works — your own captures, or a",
        "third-party dataset you convert yourself. For example, using the open chrome-fingerprints",
        "dataset (fetched by you, from its own repo, under its own licence):",
        "",
        "    pip install chrome-fingerprints",
        f'    python tools/fingerprint-collect/convert_dataset.py --out "{directory}" --count 500',
        "",
        "Or capture your own with tools/fingerprint-collect (collect.html), or point",
        "CLEARCOTE_PROFILE_SOURCE at a directory you already have.",
    ])


def _note(quiet: Optional[bool], msg: str) -> None:
    if not quiet:
        sys.stderr.write(f"[clearcote] [profile] {msg}\n")


# The optional-backup tip is worth saying once, not on every launch.
_hinted_this_process = False


def _reset_hint() -> None:
    """Test seam: reset the once-per-process tip."""
    global _hinted_this_process
    _hinted_this_process = False


def _local_index_cache_path() -> str:
    return os.path.join(profile_cache_dir(), "local-index.json")


def load_local_index(directory: str) -> list[dict[str, Any]]:
    """Index a directory, cached and invalidated by its mtime.

    Reading every file for a 10k-profile directory is seconds of I/O — far too slow to repeat per
    launch. Keying on mtime means newly converted profiles are picked up without a manual step.
    """
    mtime = os.stat(directory).st_mtime
    path = _local_index_cache_path()
    if os.path.isfile(path):
        try:
            with open(path, "r", encoding="utf-8") as f:
                c = json.load(f)
            if c.get("dir") == directory and c.get("mtime") == mtime:
                return c["index"]
        except (OSError, ValueError, KeyError):
            pass  # fall through and re-index

    index = import_directory(directory)["index"]
    try:
        os.makedirs(os.path.dirname(path), exist_ok=True)
        with open(path, "w", encoding="utf-8") as f:
            json.dump({"dir": directory, "mtime": mtime, "index": index}, f)
    except OSError:
        pass  # a cache write failure must never fail a launch
    return index


def resolve_local(host: dict[str, Any], local_dir: Optional[str] = None, **select) -> dict[str, Any]:
    """Resolve a persona from the local imported directory."""
    directory = local_dir or DEFAULT_LOCAL_DIR
    if not os.path.isdir(directory):
        raise ValueError(
            f"no local profile directory at {directory}\n\n{local_setup_hint(directory)}"
        )
    index = load_local_index(directory)
    if not index:
        raise ValueError(
            f"local profile directory {directory} contains no usable clearcote-profile JSON"
        )
    selection = select_profile(index, host, **select)
    return {
        "profile": load_imported_profile(directory, selection["entry"]["id"]),
        "selection": selection,
        "source": "local",
    }


def resolve_auto(
    host: dict[str, Any],
    license_key: Optional[str] = None,
    api_base: Optional[str] = None,
    local_dir: Optional[str] = None,
    prefer_local: bool = False,
    offline: bool = False,
    quiet: Optional[bool] = None,
    **select,
) -> dict[str, Any]:
    """Resolve ``profile="auto"``: service first, local directory as backup.

    Any service failure falls back — including 401 (no license) and 429 (rate limited or daily
    distinct budget exhausted) — because in all of those a local persona is better than none and
    far better than a hard launch failure.
    """
    global _hinted_this_process

    if prefer_local:
        r = resolve_local(host, local_dir, **select)
        _note(quiet, f"using local profile {r['selection']['entry']['id']} (prefer_local)")
        return r

    try:
        index = fetch_index(host, license_key=license_key, api_base=api_base, offline=offline,
                            os_override=select.get("os_override"))
        if not index:
            raise ValueError("service returned no candidates for this host")
        selection = select_profile(index, host, **select)
        profile = fetch_profile(selection["entry"]["id"], license_key=license_key,
                                api_base=api_base, offline=offline)
        _note(
            quiet,
            f"using service profile {selection['entry']['id']} "
            f"(score {selection['score']:.2f}, pool {selection['pool_size']}, mode {selection['mode']})",
        )
        # Mention the optional backup ONCE per process, and only while there is none. The useful
        # moment to learn about a fallback is before you need it; repeating it every launch would
        # be noise, and nagging someone who already set one up would be wrong.
        directory = local_dir or DEFAULT_LOCAL_DIR
        if not _hinted_this_process and not os.path.isdir(directory):
            _hinted_this_process = True
            _note(quiet, 'tip: no local backup directory configured. If the service is ever rate '
                         'limited or unreachable, "auto" fails rather than falling back. Setting '
                         f"one up is optional:\n{local_setup_hint(directory)}")
        return {"profile": profile, "selection": selection, "source": "service"}
    except Exception as service_err:  # noqa: BLE001 - any service failure is a fallback trigger
        reason = str(service_err)[:200]
        directory = local_dir or DEFAULT_LOCAL_DIR
        no_local_dir = not os.path.isdir(directory)
        try:
            r = resolve_local(host, local_dir, **select)
            _note(quiet, f"service unavailable ({reason}) — falling back to local profile "
                         f"{r['selection']['entry']['id']}")
            r["fallback_reason"] = reason
            return r
        except Exception as local_err:  # noqa: BLE001
            # Report BOTH failures. Surfacing only the local one would send the caller to fix a
            # directory when the real problem was an expired license.
            local_msg = f"no directory at {directory}" if no_local_dir else str(local_err)
            hint = f"\n\n{local_setup_hint(directory)}" if no_local_dir else ""
            raise ValueError(
                'profile="auto" could not resolve a persona.\n'
                f"  service: {reason}\n"
                f"  local:   {local_msg}{hint}"
            ) from local_err

File: sdk/node/src/profileauto.ts
// `profile: "auto"` — resolve a persona from the best source available.
//
// ORDER: the licensed clearcote service first, a local imported directory as backup.
//
// The service is preferred because its corpus is fresher, larger and curated. But it can be
// unreachable for reasons that are none of the caller's fault — no license, an exhausted daily
// budget, a rate limit, a network blip, an outage — and in every one of those cases a local
// profile is enormously better than no persona. So the fallback is by design, not a patch.
//
// THE FALLBACK IS ALWAYS ANNOUNCED. A silent downgrade would leave a caller believing they are
// on a fresh service profile while they are on a stale local one, and the whole product is about
// knowing exactly what identity you are presenting. `source` is returned on every result and a
// one-line reason is written to stderr unless `quiet`.

import { existsSync, mkdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import { PROFILE_DIR } from "./profile.js";
import { importDirectory, loadImportedProfile } from "./profileimport.js";
import { selectProfile, type HostFacts, type ProfileIndexEntry, type Selection } from "./profilelib.js";
import { fetchIndex, fetchProfile, type ProfileSourceOptions } from "./profilesource.js";
import { ensureLocalProfiles, hasLocalProfiles } from "./profilefetch.js";

/** Where an imported (e.g. converted chrome-fingerprints) directory is looked for by default. */
export const DEFAULT_LOCAL_DIR =
  process.env.CLEARCOTE_PROFILE_SOURCE ?? join(homedir(), ".clearcote", "profiles", "imported");

export type ProfileOrigin = "service" | "local";

/**
 * First engine major that implements `--fingerprint-profile`.
 *
 * BACKWARDS COMPATIBILITY, AND WHY THIS IS A HARD GATE. The free 149 build ships a 15-patch
 * stack with no persona-profile patch at all, so it does not merely apply the profile badly —
 * it ignores the switch entirely. Chromium discards unknown switches silently, so a 149 user
 * who asked for a profile would launch with NO persona while believing they had one. That is
 * strictly worse than the seed farbling 149 does support, because the failure is invisible.
 *
 * So a profile is never sent to an engine that cannot read it. Callers on such an engine fall
 * back to the seed path, which is what 149 was built for.
 */
export const MIN_PROFILE_ENGINE_MAJOR = 150;

/** Whether this engine can actually apply an imported profile. */
export function engineSupportsProfiles(browserMajor: number): boolean {
  return Number.isFinite(browserMajor) && browserMajor >= MIN_PROFILE_ENGINE_MAJOR;
}

export interface AutoOptions extends ProfileSourceOptions {
  /** Directory of clearcote-profile JSON used when the service is unavailable. */
  localDir?: string;
  /** Skip the service entirely and use the local directory. */
  preferLocal?: boolean;
  /** Set `false` to never bootstrap a local library on the fallback path. Default: allowed. */
  autoDownload?: boolean;
  /** How many profiles the first-run bootstrap converts (default 500). */
  autoDownloadCount?: number;
  /** Path to convert_dataset.py, for the bootstrap. Defaults to this checkout's copy, if any. */
  converter?: string;
  /** Python executable used by the bootstrap. */
  python?: string;
  quiet?: boolean;
}

export interface AutoResult {
  profile: Record<string, unknown>;
  selection: Selection;
  source: ProfileOrigin;
  /** Why the service was not used, when `source` is "local". */
  fallbackReason?: string;
}

// --- local index cache ------------------------------------------------------
// Indexing a directory means reading every file (10k profiles is a few seconds of I/O), which is
// far too slow to repeat per launch. The cache is keyed by directory and invalidated by its
// mtime, so adding or converting more profiles is picked up without a manual step.

interface CachedLocalIndex { dir: string; mtimeMs: number; index: ProfileIndexEntry[]; }

function localCachePath(): string {
  const d = join(PROFILE_DIR, "library");
  if (!existsSync(d)) mkdirSync(d, { recursive: true });
  return join(d, "local-index.json");
}

export function loadLocalIndex(dir: string): ProfileIndexEntry[] {
  const mtimeMs = statSync(dir).mtimeMs;
  const path = localCachePath();
  if (existsSync(path)) {
    try {
      const c = JSON.parse(readFileSync(path, "utf8")) as CachedLocalIndex;
      if (c.dir === dir && c.mtimeMs === mtimeMs) return c.index;
    } catch {
      /* fall through and re-index */
    }
  }
  const { index } = importDirectory(dir);
  try {
    writeFileSync(path, JSON.stringify({ dir, mtimeMs, index } satisfies CachedLocalIndex));
  } catch {
    /* a cache write failure must never fail a launch */
  }
  return index;
}

function note(quiet: boolean | undefined, msg: string): void {
  if (!quiet) process.stderr.write(`[clearcote] [profile] ${msg}\n`);
}

/** The optional-backup tip is worth saying once, not on every launch. */
let hintedThisProcess = false;

/** Test seam: reset the once-per-process tip. */
export function _resetHint(): void {
  hintedThisProcess = false;
}

/**
 * How to populate a local backup directory. Printed when the service is unavailable and no local
 * directory exists — the moment the caller actually needs it.
 *
 * OPTIONAL, AND NOT SOMETHING WE SHIP. Clearcote redistributes no third-party fingerprints; the
 * dataset is fetched by the user from its own repository under its own terms, and we only read
 * the converted format. Any directory of clearcote-profile JSON works just as well, which is
 * why this is phrased as one example rather than the required path.
 */
export function localSetupHint(dir: string): string {
  return [
    `A local profile directory is optional, and gives "auto" a backup when the service is`,
    `unavailable. Any folder of clearcote-profile JSON works — your own captures, or a`,
    `third-party dataset you convert yourself. For example, using the open chrome-fingerprints`,
    `dataset (fetched by you, from its own repo, under its own licence):`,
    ``,
    `    pip install chrome-fingerprints`,
    `    python tools/fingerprint-collect/convert_dataset.py --out "${dir}" --count 500`,
    ``,
    `Or capture your own with tools/fingerprint-collect (collect.html), or point`,
    `CLEARCOTE_PROFILE_SOURCE at a directory you already have.`,
  ].join("\n");
}

/** Resolve a persona from the local imported directory. */
export function resolveLocal(host: HostFacts, opts: AutoOptions = {}): AutoResult {
  const dir = opts.localDir ?? DEFAULT_LOCAL_DIR;
  if (!existsSync(dir)) {
    throw new Error(`no local profile directory at ${dir}\n\n${localSetupHint(dir)}`);
  }
  const index = loadLocalIndex(dir);
  if (index.length === 0) {
    throw new Error(`local profile directory ${dir} contains no usable clearcote-profile JSON`);
  }
  const selection = selectProfile(index, host, opts);
  return { profile: loadImportedProfile(dir, selection.entry.id), selection, source: "local" };
}

/**
 * Resolve `profile: "auto"`: service first, local directory as backup.
 *
 * Any service failure falls back — including 401 (no license) and 429 (rate limited or daily
 * distinct budget exhausted), because in all of those a local persona is better than none and
 * far better than a hard launch failure.
 */
export async function resolveAuto(host: HostFacts, opts: AutoOptions = {}): Promise<AutoResult> {
  if (opts.preferLocal) {
    const r = resolveLocal(host, opts);
    note(opts.quiet, `using local profile ${r.selection.entry.id} (preferLocal)`);
    return r;
  }

  try {
    const index = await fetchIndex(host, opts);
    if (index.length === 0) throw new Error("service returned no candidates for this host");
    const selection = selectProfile(index, host, opts);
    const profile = await fetchProfile(selection.entry.id, opts);
    note(
      opts.quiet,
      `using service profile ${selection.entry.id} ` +
        `(score ${selection.score.toFixed(2)}, pool ${selection.poolSize}, mode ${selection.mode})`,
    );
    // Mention the optional backup ONCE per process, and only while there is none configured.
    // The useful moment to learn about a fallback is before you need it — but repeating it on
    // every launch would be noise, and nagging someone who already set one up would be wrong.
    if (!hintedThisProcess && !existsSync(opts.localDir ?? DEFAULT_LOCAL_DIR)) {
      hintedThisProcess = true;
      note(opts.quiet, `tip: no local backup directory configured. If the service is ever rate limited or
unreachable, "auto" fails rather than falling back. Setting one up is optional:
${localSetupHint(opts.localDir ?? DEFAULT_LOCAL_DIR)}`);
    }
    return { profile, selection, source: "service" };
  } catch (serviceErr) {
    const reason = String(serviceErr instanceof Error ? serviceErr.message : serviceErr).slice(0, 200);
    // First-run bootstrap: if there is no local library yet, try to build one before giving up.
    // Only on the fallback path — a working service must never pay for this. Never throws: a
    // bootstrap failure has to degrade to "no local backup", not take down the launch.
    if (opts.autoDownload !== false) {
      const dir = opts.localDir ?? DEFAULT_LOCAL_DIR;
      if (!hasLocalProfiles(dir)) {
        await ensureLocalProfiles(dir, join(PROFILE_DIR, "library"), {
          quiet: opts.quiet,
          count: opts.autoDownloadCount,
          converter: opts.converter,
          python: opts.python,
        }).catch(() => false);
      }
    }
    try {
      const r = resolveLocal(host, opts);
      // Both halves of the story: what failed, and what is being used instead.
      note(opts.quiet, `service unavailable (${reason}) — falling back to local profile ${r.selection.entry.id}`);
      return { ...r, fallbackReason: reason };
    } catch (localErr) {
      const dir = opts.localDir ?? DEFAULT_LOCAL_DIR;
      const noLocalDir = !existsSync(dir);
      const lr = String(localErr instanceof Error ? localErr.message : localErr);
      // Report BOTH failures. Surfacing only the local one would send the caller to go and fix
      // a directory when the real problem was an expired license.
      throw new Error(
        `profile: "auto" could not resolve a persona.\n` +
          `  service: ${reason}\n` +
          `  local:   ${noLocalDir ? `no directory at ${dir}` : lr}\n\n` +
          (noLocalDir ? localSetupHint(dir) : ""),
      );
    }
  }
}
